3.8.0: unit-cbor.cpp test failures

Hi! I package nlohmann-json for Arch Linux. Upon building version 3.8.0 I ran the unit tests, but some fail.

For the last stable release (3.7.3) I used the below described workflow and the unit tests pass. Maybe something has changed in the meantime?

#### What is the issue you have?

```
Executing test-algorithms...
[doctest] doctest version is "2.3.7"
[doctest] run with "--help" for options
===============================================================================
[doctest] test cases:      1 |      1 passed |      0 failed |      0 skipped
[doctest] assertions:     39 |     39 passed |      0 failed |
[doctest] Status: SUCCESS!
Executing test-allocator...
[doctest] doctest version is "2.3.7"
[doctest] run with "--help" for options
===============================================================================
[doctest] test cases:      3 |      3 passed |      0 failed |      0 skipped
[doctest] assertions:     17 |     17 passed |      0 failed |
[doctest] Status: SUCCESS!
Executing test-alt-string...
[doctest] doctest version is "2.3.7"
[doctest] run with "--help" for options
===============================================================================
[doctest] test cases:      1 |      1 passed |      0 failed |      0 skipped
[doctest] assertions:     28 |     28 passed |      0 failed |
[doctest] Status: SUCCESS!
Executing test-bson...
[doctest] doctest version is "2.3.7"
[doctest] run with "--help" for options
===============================================================================
[doctest] test cases:      6 |      6 passed |      0 failed |      1 skipped
[doctest] assertions:    486 |    486 passed |      0 failed |
[doctest] Status: SUCCESS!
Executing test-capacity...
[doctest] doctest version is "2.3.7"
[doctest] run with "--help" for options
===============================================================================
[doctest] test cases:      1 |      1 passed |      0 failed |      0 skipped
[doctest] assertions:    120 |    120 passed |      0 failed |
[doctest] Status: SUCCESS!
Executing test-cbor...
[doctest] doctest version is "2.3.7"
[doctest] run with "--help" for options
===============================================================================
src/unit-cbor.cpp:1912:
TEST CASE:  single CBOR roundtrip
  sample.json

src/unit-cbor.cpp:1912: ERROR: test case THREW exception: exception thrown in subcase - will translate later when the whole test case has been exited (cannot translate while there is an active exception)

===============================================================================
src/unit-cbor.cpp:1912:
TEST CASE:  single CBOR roundtrip

src/unit-cbor.cpp:1912: ERROR: test case THREW exception: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal

===============================================================================
src/unit-cbor.cpp:2350:
TEST CASE:  examples from RFC 7049 Appendix A
  byte arrays

src/unit-cbor.cpp:2500: ERROR: CHECK_NOTHROW( j = json::from_cbor(packed) ) THREW exception: "[json.exception.parse_error.110] parse error at byte 1: syntax error while parsing CBOR value: unexpected end of input"

src/unit-cbor.cpp:2505: ERROR: CHECK( j == json::binary(expected) ) is NOT correct!
  values: CHECK( null == {"bytes":[],"subtype":null} )

===============================================================================
[doctest] test cases:      5 |      3 passed |      2 failed |      1 skipped
[doctest] assertions: 1640832 | 1640830 passed |      2 failed |
[doctest] Status: FAILURE!
```

Full build log: 
[nlohmann-json-3.8.0-build.log](https://github.com/nlohmann/json/files/4776937/nlohmann-json-3.8.0-build.log)

#### Please describe the steps to reproduce the issue.

I'm cloning this repository from the commit (as no annotated tags exist) and verify the lightweight tag/commit using the author's PGP signature.

I'm building with:

```
  cmake -DCMAKE_INSTALL_PREFIX='/usr' \
        -DCMAKE_INSTALL_LIBDIR='/usr/lib' \
        -DCMAKE_BUILD_TYPE='None' \
        -DBUILD_TESTING=ON \
        -DJSON_MultipleHeaders=ON \
        -Wno-dev \
        -B build \
        -S .
  make VERBOSE=1 -C build
```

Additional to the above I run the following to get pkgconfig integration (but that has nothing to do with the tests):

```
  meson --prefix=/usr \
        --libdir=lib \
        meson-build
  ninja -C meson-build
```

I run tests with:
```
make -k check
```

Note:
When running the test suite using ctest (e.g. as in use in the travis integration)

https://github.com/nlohmann/json/blob/e7452d87783fbf6e9d320d515675e26dfd1271c5/.travis.yml#L324-L328

No tests are collected:

```
UpdateCTestConfiguration  from :/build/nlohmann-json/src/nlohmann-json-3.8.0/DartConfiguration.tcl
UpdateCTestConfiguration  from :/build/nlohmann-json/src/nlohmann-json-3.8.0/DartConfiguration.tcl
Test project /build/nlohmann-json/src/nlohmann-json-3.8.0
Constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
No tests were found!!!
```

#### Can you provide a small but working code example?

n/a

#### What is the expected behavior?

All unit-tests pass

#### And what is the actual behavior instead?

Some `unit-cbor` unit tests fail.

#### Which compiler and operating system are you using?

- Compiler: gcc 10.1.0
- Operating system: Arch Linux

#### Which version of the library did you use?

- [ ] latest release version 3.7.3
- [x] other release - please state the version: 3.8.0
- [ ] the `develop` branch

#### If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

- [ ] yes
- [x] no - please copy/paste the error message below

see above

> Can you provide the output of the failed tests?

```text
50/51 Test #50: cmake_fetch_content_configure ......***Failed    0.36 sec
-- The CXX compiler identification is GNU 10.1.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
[1/9] Creating directories for 'json-populate'
[1/9] Performing download step (git clone) for 'json-populate'
fatal: repository '/builddir/build/BUILD/json-3.8.0/test/cmake_fetch_content/project/../../..' does not exist
fatal: repository '/builddir/build/BUILD/json-3.8.0/test/cmake_fetch_content/project/../../..' does not exist
fatal: repository '/builddir/build/BUILD/json-3.8.0/test/cmake_fetch_content/project/../../..' does not exist
-- Had to git clone more than once:
          3 times.
CMake Error at json-subbuild/json-populate-prefix/tmp/json-populate-gitclone.cmake:31 (message):
  Failed to clone repository:
  '/builddir/build/BUILD/json-3.8.0/test/cmake_fetch_content/project/../../..'


FAILED: json-populate-prefix/src/json-populate-stamp/json-populate-download 
cd /builddir/build/BUILD/json-3.8.0/x86_64-redhat-linux-gnu/test/cmake_fetch_content/_deps && /usr/bin/cmake -P /builddir/build/BUILD/json-3.8.0/x86_64-redhat-linux-gnu/test/cmake_fetch_content/_deps/json-subbuild/json-populate-prefix/tmp/json-populate-gitclone.cmake && /usr/bin/cmake -E touch /builddir/build/BUILD/json-3.8.0/x86_64-redhat-linux-gnu/test/cmake_fetch_content/_deps/json-subbuild/json-populate-prefix/src/json-populate-stamp/json-populate-download
ninja: build stopped: subcommand failed.

CMake Error at /usr/share/cmake/Modules/FetchContent.cmake:912 (message):
  Build step for json failed: 1
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FetchContent.cmake:1003 (__FetchContent_directPopulate)
  CMakeLists.txt:13 (FetchContent_Populate)


-- Configuring incomplete, errors occurred!
See also "/builddir/build/BUILD/json-3.8.0/x86_64-redhat-linux-gnu/test/cmake_fetch_content/CMakeFiles/CMakeOutput.log".

      Start 51: cmake_fetch_content_build
Failed test dependencies: cmake_fetch_content_configure
51/51 Test #51: cmake_fetch_content_build ..........***Not Run   0.00 sec

96% tests passed, 2 tests failed out of 51
```

> Which CMake version are you using?

cmake-3.17.3-1.fc32.x86_64

> moves the find_package(Git) and tries to fix the cmake_fetch_content_configure test. Please check.

Now it builds fine without installed Git. Thanks.

But tests cmake_fetch_content_configure and cmake_fetch_content_build are still broken due to missing Git and absent Internet access. New logs:

```text
50/51 Test #50: cmake_fetch_content_configure ......***Failed    0.34 sec
-- The CXX compiler identification is GNU 10.1.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at /usr/share/cmake/Modules/ExternalProject.cmake:2454 (message):
  error: could not find git for clone of json-populate
Call Stack (most recent call first):
  /usr/share/cmake/Modules/ExternalProject.cmake:3271 (_ep_add_download_command)
  CMakeLists.txt:13 (ExternalProject_Add)


-- Configuring incomplete, errors occurred!
See also "/builddir/build/BUILD/json-3.8.0/x86_64-redhat-linux-gnu/test/cmake_fetch_content/_deps/json-subbuild/CMakeFiles/CMakeOutput.log".

CMake Error at /usr/share/cmake/Modules/FetchContent.cmake:900 (message):
  CMake step for json failed: 1
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FetchContent.cmake:1003 (__FetchContent_directPopulate)
  CMakeLists.txt:12 (FetchContent_Populate)


-- Configuring incomplete, errors occurred!
See also "/builddir/build/BUILD/json-3.8.0/x86_64-redhat-linux-gnu/test/cmake_fetch_content/CMakeFiles/CMakeOutput.log".

      Start 51: cmake_fetch_content_build
Failed test dependencies: cmake_fetch_content_configure
51/51 Test #51: cmake_fetch_content_build ..........***Not Run   0.00 sec

96% tests passed, 2 tests failed out of 51
```

@nlohmann Just disable these tests if the `JSON_TestDataDirectory` variable is set.

--- a/README.md
+++ b/README.md
@@ -1557,4 +1557,6 @@ $ cmake --build .
 $ ctest --output-on-failure
 ```
 
-For more information, have a look at the file [.travis.yml](https://github.com/nlohmann/json/blob/master/.travis.yml).
+Note that during the `ctest` stage, several JSON test files are downloaded from an [external repository](https://github.com/nlohmann/json_test_data). If policies forbid downloading artifacts during testing, you can download the files yourself and pass the directory with the test files via `-DJSON_TestDataDirectory=path` to CMake. Then, no Internet connectivity is required. See [issue #2189](https://github.com/nlohmann/json/issues/2189) for more information.
+
+In case you have downloaded the library rather than checked out the code via Git, test `cmake_fetch_content_configure`. Please execute `ctest -LE git_required` to skip these tests. See [issue #2189](https://github.com/nlohmann/json/issues/2189) for more information.

--- a/cmake/download_test_data.cmake
+++ b/cmake/download_test_data.cmake
@@ -1,5 +1,3 @@
-find_package(Git)
-
 set(JSON_TEST_DATA_URL     https://github.com/nlohmann/json_test_data)
 set(JSON_TEST_DATA_VERSION 2.0.0)
 
@@ -9,6 +7,7 @@ if(JSON_TestDataDirectory)
     add_custom_target(download_test_data)
     file(WRITE ${CMAKE_BINARY_DIR}/include/test_data.hpp "#define TEST_DATA_DIRECTORY \"${JSON_TestDataDirectory}\"\n")
 else()
+    find_package(Git)
     # target to download test data
     add_custom_target(download_test_data
         COMMAND test -d json_test_data || ${GIT_EXECUTABLE} clone -c advice.detachedHead=false --branch v${JSON_TEST_DATA_VERSION} ${JSON_TEST_DATA_URL}.git --quiet --depth 1

--- a/cmake/download_test_data.cmake
+++ b/cmake/download_test_data.cmake
@@ -3,15 +3,21 @@ find_package(Git)
 set(JSON_TEST_DATA_URL     https://github.com/nlohmann/json_test_data)
 set(JSON_TEST_DATA_VERSION 2.0.0)
 
-# target to download test data
-add_custom_target(download_test_data
-    COMMAND test -d json_test_data || ${GIT_EXECUTABLE} clone -c advice.detachedHead=false --branch v${JSON_TEST_DATA_VERSION} ${JSON_TEST_DATA_URL}.git --quiet --depth 1
-    COMMENT "Downloading test data from ${JSON_TEST_DATA_URL} (v${JSON_TEST_DATA_VERSION})"
-    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
-)
-
-# create a header with the path to the downloaded test data
-file(WRITE ${CMAKE_BINARY_DIR}/include/test_data.hpp "#define TEST_DATA_DIRECTORY \"${CMAKE_BINARY_DIR}/json_test_data\"\n")
+# if variable is set, use test data from given directory rather than downloading them
+if(JSON_TestDataDirectory)
+    message(STATUS "Using test data in ${JSON_TestDataDirectory}.")
+    add_custom_target(download_test_data)
+    file(WRITE ${CMAKE_BINARY_DIR}/include/test_data.hpp "#define TEST_DATA_DIRECTORY \"${JSON_TestDataDirectory}\"\n")
+else()
+    # target to download test data
+    add_custom_target(download_test_data
+        COMMAND test -d json_test_data || ${GIT_EXECUTABLE} clone -c advice.detachedHead=false --branch v${JSON_TEST_DATA_VERSION} ${JSON_TEST_DATA_URL}.git --quiet --depth 1
+        COMMENT "Downloading test data from ${JSON_TEST_DATA_URL} (v${JSON_TEST_DATA_VERSION})"
+        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+    )
+    # create a header with the path to the downloaded test data
+    file(WRITE ${CMAKE_BINARY_DIR}/include/test_data.hpp "#define TEST_DATA_DIRECTORY \"${CMAKE_BINARY_DIR}/json_test_data\"\n")
+endif()
 
 # determine the operating system (for debug and support purposes)
 find_program(UNAME_COMMAND uname)

--- a/test/cmake_fetch_content/CMakeLists.txt
+++ b/test/cmake_fetch_content/CMakeLists.txt
@@ -11,8 +11,10 @@ if (${CMAKE_VERSION} VERSION_GREATER "3.11.0")
   )
   set_tests_properties(cmake_fetch_content_configure PROPERTIES
     FIXTURES_SETUP cmake_fetch_content
+    LABELS git_required
   )
   set_tests_properties(cmake_fetch_content_build PROPERTIES
     FIXTURES_REQUIRED cmake_fetch_content
+    LABELS git_required
   )
 endif()

--- a/test/cmake_fetch_content/project/CMakeLists.txt
+++ b/test/cmake_fetch_content/project/CMakeLists.txt
@@ -4,9 +4,8 @@ project(DummyImport CXX)
 
 include(FetchContent)
 
-FetchContent_Declare(json
-  GIT_REPOSITORY ${CMAKE_CURRENT_SOURCE_DIR}/../../..
-  GIT_TAG HEAD)
+get_filename_component(GIT_REPOSITORY_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/../../.. ABSOLUTE)
+FetchContent_Declare(json GIT_REPOSITORY ${GIT_REPOSITORY_DIRECTORY} GIT_TAG HEAD)
 
 FetchContent_GetProperties(json)
 if(NOT json_POPULATED)

